### PR TITLE
feat(grpc-gateway) encode Timestamp

### PR DESCRIPTION
**Pull request by javierguerragiraldez**
Transform the "well known type" `.google.protobuf.Timestamp` (a
secs+nanos protobuf structure) to/from JSON datetime string.

requires the `encode hooks` feature starwing/lua-protobuf#145

### CODE REVIEW

**Comment by fffonion** [comment]
👍 @javierguerragiraldez Do we also want to include cases described in https://github.com/Kong/kong/issues/7469?

**Comment by dndx** [comment]
Hey @javierguerragiraldez , do we need any more tests for this as @fffonion suggested?

**Comment by javierguerragiraldez** [comment]
Yes, that's what the new grpc target is for. Won't run in IC without the Google includes, though.

**Review by dndx — APPROVED** [review]

File: kong/plugins/grpc-gateway/deco.lua [hunk 1]
```diff
@@ -78,6 +79,47 @@ local function parse_options_path(path)
...
 end
 
+local function safe_set_type_hook(type, dec, enc)
+  if not pcall(pb.hook, type) then
+    ngx.log(ngx.NOTICE, "no type '" .. type .. "' defined")
+    return
+  end
+
+  if not pb.hook(type) then
+    pb.hook(type, dec)
```
Comment (git-torrent) [inline kong/plugins/grpc-gateway/deco.lua:89]: Sadly, this solution cannot handle native types, namely `bytes` - https://github.com/starwing/lua-protobuf/issues/161#issuecomment-927001222

Bytes are handled in special way https://developers.google.com/protocol-buffers/docs/proto3#json

We fixed this by manually traversing the message looking for type and calling respective hook. I can push my solution, but first I want to integrate it with yours.

**Pull request merged**